src/Random/InverseTransformSampling.h

```diff
@@ -0,0 +1,173 @@
+                umax[d] = dist.cdf(regions(rank)[d].max(),d);
+            }
+            T pnr = std::accumulate(nr_m.begin(), nr_m.end(), 1.0, std::multiplies<T>());
+            T pdr = std::accumulate(dr_m.begin(), dr_m.end(), 1.0, std::multiplies<T>());
+            
+            T factor = pnr / pdr;
+            nlocal_m      = factor * ntotal;
+            
+            unsigned int ngobal = 0;
+            MPI_Allreduce(&nlocal_m, &ngobal, 1, MPI_UNSIGNED_LONG, MPI_SUM,
+                          ippl::Comm->getCommunicator());
+            
+            int rest = (int)(ntotal - ngobal);
+            
+            if (rank < rest) {
+                ++nlocal_m;
+            }
+        }
+
+        /*!
+         * @brief Functor that is used for generating samples.
+        */
+        template <class GeneratorPool>
+        struct fill_random {
+            using value_type = T;
+            Distribution dist;
+            view_type x;
+            GeneratorPool rand_pool;
+            Vector<T, Dim> umin_m;
+            Vector<T, Dim> umax_m;
+            
+            /*!
+             * @brief Constructor for the fill_random functor.
+             *
+             * @param dist_ The distribution to sample from.
+             * @param x_ The view to generate samples in.
+             * @param rand_pool_ The random number generator pool.
+             * @param umin_ Minimum cumulative distribution values.
+             * @param umax_ Maximum cumulative distribution values.
+            */
+            KOKKOS_FUNCTION
+            fill_random(Distribution dist_, view_type x_, GeneratorPool rand_pool_, Vector<T, Dim> umin_, Vector<T, Dim> umax_)
+            : dist(dist_)
+            , x(x_)
+            , rand_pool(rand_pool_)
+            , umin_m(umin_)
+            , umax_m(umax_){}
+            
+            /*!
+             * @brief Operator to fill random values.
+             *
+             * @param i Index for the random values.
+            */
+            KOKKOS_INLINE_FUNCTION void operator()(const size_t i) const {
+                typename GeneratorPool::generator_type rand_gen = rand_pool.get_state();
+                
+                value_type u = 0.0;
+                for (unsigned d = 0; d < Dim; ++d) {
+                    u       = rand_gen.drand(umin_m[d], umax_m[d]);
+                    
+                    // first guess for Newton-Raphson
+                    x(i)[d] = dist.estimate(u, d);
+                    
+                    // solve
+                    ippl::random::detail::NewtonRaphson<value_type> solver;
+
+                    solver.solve(dist, d, x(i)[d], u);
+
+                    rand_pool.free_state(rand_gen);
+                }
+            }
+        };
+        
+        /*!
+         * @brief Get the local number of samples.
+         *
+         * @returns The local number of samples.
+        */
+        KOKKOS_INLINE_FUNCTION unsigned int getLocalNum() const { return nlocal_m; }
```

Comment:
getLocalNum is computed in this class which can be used for generating samples per rank. The local number of particles in ParticleBase can be set using this function.

Reply:
As far as I know we don't set explicitly (apart from during communication etc.) the local number of particles in ParticleBase. They should come automatically from `create` particles.

Reply:
This could be useful in case the user wants to generate samples of a distribution without using ParticleBase class.

Reply:
In that case it may be better to change the name to "getLocalSamplesNum()" or something similar so that at the user level it does not get confused with the "getLocalNum()" (which is what should be used during evolution of particles) from ParticleBase.

Reply:
done in 3db4cde

Reply:
It is strange that it doesn't show here the commit as you then don't need to explicitly mention like "done in...". Gitlab used to show it like that and I thought github would also do the same but doesn't seem like it and it shows the commits only at the end.